### ISSUE TITLE
Finetune ImageReward?

### ISSUE THREAD

**Issue by hkunzhe** [THUDM/ImageReward#45]
Hi, I want to use my personal preference ranking data to fine-tune ImageReward. But I have encountered several problems:
1. Why use [BertTokenizer](https://github.com/THUDM/ImageReward/blob/main/train/src/rank_pair_dataset.py#L40) rather than `blip.text_encoder` when training reward models?
2. Given preference data (including train, validation and test) in [test.json](https://github.com/THUDM/ImageReward/blob/main/data/test.json) format, can I use [encode_data](https://github.com/THUDM/ImageReward/blob/main/train/src/ImageReward.py#L121) in `forward` to compute the ranking loss rather than [make_dataset](https://github.com/THUDM/ImageReward/blob/main/train/src/make_dataset.py) first?

Thank you in advance!

**Comment by 1073521013** [comment]
Very good work, please ask a reward training problem, I used the code and hyperparameters provided by the paper, the result is that whether you provide data or my own data, fine-tune or train from scratch, the loss is always 0.69315, and the reward value is always -0.16178 when testing, Please help see what's going on

**Comment by 1073521013** [comment]
> 1. They are the same. Refer to [blip.py](https://github.com/THUDM/ImageReward/blob/main/train/src/models/blip.py#L17C13-L17C13).
> 2. Sure, you can do that. make_dataset is to save on calculations.

Very good work, please ask a reward training problem, I used the code and hyperparameters provided by the paper, the result is that whether you provide data or my own data, fine-tune or train from scratch, the loss is always 0.69315, and the reward value is always -0.16178 when testing, Please help see what's going on

**Comment by hkunzhe** [comment]
> > 1. They are the same. Refer to [blip.py](https://github.com/THUDM/ImageReward/blob/main/train/src/models/blip.py#L17C13-L17C13).
> > 2. Sure, you can do that. make_dataset is to save on calculations.
> 
> Very good work, please ask a reward training problem, I used the code and hyperparameters provided by the paper, the result is that whether you provide data or my own data, fine-tune or train from scratch, the loss is always 0.69315, and the reward value is always -0.16178 when testing, Please help see what's going on

I have met the same weird problem. So I reopen this issue.

**Comment by choi403** [comment]
Hi, I'm also experiencing the same issue. The reward value of a fine-tuned ImageReward model for any image and text is exactly -0.1617799699306488.

**Comment by choi403** [comment]
Hi, I think I've figured out the problem.
```python
model = ImageReward(device='cuda', med_config='/home/username/.cache/ImageReward/med_config.json').to('cuda')

# Fine-tuned ImageReward 
my_state_dict = torch.load('my_image_reward_model.pt', map_location='cpu')
print(my_state_dict)
# OrderedDict([('module.blip.visual_encoder.cls_token', tensor([[[ 0.1317,  0.0138, -0.0130,  ...,  0.0264, -0.0023,  0.0311]]])), ('module.blip.visual_encoder.pos_embed', tensor([[[ 0.1333,  0.0110, -0.0103,  ...,  0.0295, -0.0059,  0.0355],
#          [ 0.0679,  0.0161, -0.0033,  ...,  0.0028,  0.0045,  0.0154],
#          [ 0.0833,  0.0077, -0.0069,  ...,  0.0154, -0.0028, -0.0060],
#          ...,
#          [ 0.0198, -0.0029, -0.0020,  ...,  0.0234, -0.0102,  0.0307],
#          [ 0.0110, -0.0065,  0.0224,  ..., -0.0079, -0.0072,  0.0165],
#          [ 0.0241, -0.0008, -0.0054,  ...,  0.0096, -0.0043,  0.0250]]])), ...

# Vanilla ImageReward 
vanilla_state_dict = torch.load('/home/username/.cache/ImageReward/ImageReward.pt', map_location='cpu')
print(vanilla_state_dict)
# {'blip.visual_encoder.cls_token': tensor([[[ 0.1317,  0.0138, -0.0130,  ...,  0.0264, -0.0023,  0.0311]]]), 'blip.visual_encoder.pos_embed': tensor([[[ 0.1333,  0.0110, -0.0103,  ...,  0.0295, -0.0059,  0.0355],
#          [ 0.0679,  0.0161, -0.0033,  ...,  0.0028,  0.0045,  0.0154],
#          [ 0.0833,  0.0077, -0.0069,  ...,  0.0154, -0.0028, -0.0060],
#          ...,
#          [ 0.0198, -0.0029, -0.0020,  ...,  0.0234, -0.0102,  0.0307],
#          [ 0.0110, -0.0065,  0.0224,  ..., -0.0079, -0.0072,  0.0165],
#          [ 0.0241, -0.0008, -0.0054,  ...,  0.0096, -0.0043,  0.0250]]]), ...

# Convert state_dict to dict, remove prefix "module."
my_state_dict = {key.split('module.')[1]: value for key, value in my_state_dict.items()}

# Load and use
model.load_state_dict(my_state_dict,strict=False)
with torch.no_grad():
    reward = model.score("prompt text", "test.png")
```
It seems that the save format of the provided fine-tuning code is OrderedDict, while the vanilla ImageReward checkpoint is dict. I've converted it to dict, and removed the "module." prefix that was added because of multi gpu training, and it seems to be working now.
```python
model = ImageReward(device='cuda', med_config='/home/username/.cache/ImageReward/med_config.json').to('cuda')
my_state_dict = torch.load('my_image_reward_model.pt', map_location='cpu')
my_state_dict = {key.split('module.')[1]: value for key, value in my_state_dict.items()}
model.load_state_dict(my_state_dict,strict=False)
```

**Comment by 1073521013** [comment]
But I still feel it's difficult to converge when training from scratch. Did you have any effect on fine-tuning or training the reward model from scratch with your own data

> Hi, I think I've figured out the problem.
> 
> ```python
> model = ImageReward(device='cuda', med_config='/home/username/.cache/ImageReward/med_config.json').to('cuda')
> 
> # Fine-tuned ImageReward 
> my_state_dict = torch.load('my_image_reward_model.pt', map_location='cpu')
> print(my_state_dict)
> # OrderedDict([('module.blip.visual_encoder.cls_token', tensor([[[ 0.1317,  0.0138, -0.0130,  ...,  0.0264, -0.0023,  0.0311]]])), ('module.blip.visual_encoder.pos_embed', tensor([[[ 0.1333,  0.0110, -0.0103,  ...,  0.0295, -0.0059,  0.0355],
> #          [ 0.0679,  0.0161, -0.0033,  ...,  0.0028,  0.0045,  0.0154],
> #          [ 0.0833,  0.0077, -0.0069,  ...,  0.0154, -0.0028, -0.0060],
> #          ...,
> #          [ 0.0198, -0.0029, -0.0020,  ...,  0.0234, -0.0102,  0.0307],
> #          [ 0.0110, -0.0065,  0.0224,  ..., -0.0079, -0.0072,  0.0165],
> #          [ 0.0241, -0.0008, -0.0054,  ...,  0.0096, -0.0043,  0.0250]]])), ...
> 
> # Vanilla ImageReward 
> vanilla_state_dict = torch.load('/home/username/.cache/ImageReward/ImageReward.pt', map_location='cpu')
> print(vanilla_state_dict)
> # {'blip.visual_encoder.cls_token': tensor([[[ 0.1317,  0.0138, -0.0130,  ...,  0.0264, -0.0023,  0.0311]]]), 'blip.visual_encoder.pos_embed': tensor([[[ 0.1333,  0.0110, -0.0103,  ...,  0.0295, -0.0059,  0.0355],
> #          [ 0.0679,  0.0161, -0.0033,  ...,  0.0028,  0.0045,  0.0154],
> #          [ 0.0833,  0.0077, -0.0069,  ...,  0.0154, -0.0028, -0.0060],
> #          ...,
> #          [ 0.0198, -0.0029, -0.0020,  ...,  0.0234, -0.0102,  0.0307],
> #          [ 0.0110, -0.0065,  0.0224,  ..., -0.0079, -0.0072,  0.0165],
> #          [ 0.0241, -0.0008, -0.0054,  ...,  0.0096, -0.0043,  0.0250]]]), ...
> 
> # Convert state_dict to dict, remove prefix "module."
> my_state_dict = {key.split('module.')[1]: value for key, value in my_state_dict.items()}
> 
> # Load and use
> model.load_state_dict(my_state_dict,strict=False)
> with torch.no_grad():
>     reward = model.score("prompt text", "test.png")
> ```
> 
> It seems that the save format of the provided fine-tuning code is OrderedDict, while the vanilla ImageReward checkpoint is dict. I've converted it to dict, and removed the "module." prefix that was added because of multi gpu training, and it seems to be working now.
> 
> ```python
> model = ImageReward(device='cuda', med_config='/home/username/.cache/ImageReward/med_config.json').to('cuda')
> my_state_dict = torch.load('my_image_reward_model.pt', map_location='cpu')
> my_state_dict = {key.split('module.')[1]: value for key, value in my_state_dict.items()}
> model.load_state_dict(my_state_dict,strict=False)
> ```

**Comment by choi403** [comment]
I've only tried fine-tuning the model on a custom dataset and continuing training from the ImageReward v1.0 checkpoint, and it seems that it is working at least for me, as the test accuracy for my custom dataset is increasing and the accuracy for ImageReward dataset is slightly decreasing, which I believe is to be expected when you're fine-tuning on a different dataset. Maybe you could try fine-tuning it instead of training from scratch? If I recall correctly, I'm also using the same hyperparameters as written by the paper authors.

**Comment by hkunzhe** [comment]
@choi403 ,what's the size of your custom dataset?

**Comment by xujz18** [comment]
Thanks a lot to @choi403 for a valuable discussion! Fine-tuning from the ImageReward v1.0 checkpoint will be more effective.
Two points may be help:
1. The training process exhibits sensitivity to hyper-parameters such as the fix rate, learning rate, and batch size. You can consider modifying these hyper-parameters to enhance training procedure.
2. It is acknowledged that during the initial stages of training, there may be instances where the loss remains temporarily unchanged. You can extend the training duration to observe if this issue resolves itself.

**Comment by choi403** [comment]
Hi @hkunzhe, I'm using a dataset consisting of roughly 20,000 image pairs. Test accuracy on the custom dataset seems to converge at roughly ~4 epochs, in my case.

**Comment by hkunzhe** [comment]
> Hi @hkunzhe, I'm using a dataset consisting of roughly 20,000 image pairs. Test accuracy on the custom dataset seems to converge at roughly ~4 epochs, in my case.

Thanks a lot!
